### PR TITLE
fix: do not render labels that are located outside of parent rect

### PR DESCRIPTION
**Pull request by nguyenlam1**
- If the label(s) are located totally or partially outside of parent rect, it is not rendered.

**Checklist**

- [x] tests added
- [x] commits conform to the [commit guidelines](./CONTRIBUTING.md#commit)
- [ ] documentation updated

### CODE REVIEW

File: packages/picasso.js/src/core/chart-components/labels/strategies/slices.js [hunk 1]
```diff
@@ -8,6 +8,20 @@ function pad(bounds, padding) {
   bounds.height -= (padding * 2);
 }
 
+/**
+ * Check if the pie slice label is located entirely within it's parent rect
+ * @param {Object} bounds - An object describing the position(x,y), width and height of label
```
Comment (miralemd) [inline packages/picasso.js/src/core/chart-components/labels/strategies/slices.js]: both parameters should be of type `rect`

Comment (nguyenlam1) [reply packages/picasso.js/src/core/chart-components/labels/strategies/slices.js]: so it should be @param {rect} bounds?

Comment (miralemd) [reply packages/picasso.js/src/core/chart-components/labels/strategies/slices.js]: Yes

Comment (nguyenlam1) [reply packages/picasso.js/src/core/chart-components/labels/strategies/slices.js]: fixed!

File: packages/picasso.js/src/core/chart-components/labels/strategies/slices.js [hunk 2]
```diff
@@ -8,6 +8,20 @@ function pad(bounds, padding) {
...
+ * @param {Object} rect - An object describing the position(x,y), width and height of parent rect
+ * @ignore
+ */
+function isInRect(bounds, rect) {
```
[the remark this answers is not in the document]
Comment (nguyenlam1) [reply packages/picasso.js/src/core/chart-components/labels/strategies/slices.js]: fixed!

File: packages/picasso.js/src/core/math/intersection.js [hunk 3]
```diff
@@ -17,3 +17,25 @@ export function isCollinear(p1, p2, p3, e = EPSILON) {
...
+ * @param {rect} childRect - A rect describing the position(x,y), width and height of label
+ * @param {rect} parentRect - A rect describing the position(x,y), width and height of parent rect
+ * @ignore
+ */
+export function rectContainsRect(childRect, parentRect) {
+  if (!childRect || !parentRect) {
+    throw new Error('A rect or both rects are missing');
+  }
+
+  if (!childRect.x || !childRect.y || !parentRect.width || !parentRect.height) {
```
Comment (cbt1) [inline packages/picasso.js/src/core/math/intersection.js]: This wont work if any of these properties are zero

File: packages/picasso.js/src/core/math/intersection.js [hunk 4]
```diff
@@ -17,3 +17,25 @@ export function isCollinear(p1, p2, p3, e = EPSILON) {
   const t = Math.abs((p1.x * (p2.y - p3.y)) + (p2.x * (p3.y - p1.y)) + (p3.x * (p1.y - p2.y)));
   return t < e;
 }
+
+/**
+ * Check if the pie slice label is located entirely within it's parent rect
```
Comment (cbt1) [inline packages/picasso.js/src/core/math/intersection.js]: Remove pie slice reference and add some description for the function

File: packages/picasso.js/src/core/math/intersection.js [hunk 5]
```diff
@@ -17,3 +17,25 @@ export function isCollinear(p1, p2, p3, e = EPSILON) {
...
+ * @param {rect} childRect - A rect describing the position(x,y), width and height of label
+ * @param {rect} parentRect - A rect describing the position(x,y), width and height of parent rect
+ * @ignore
+ */
+export function rectContainsRect(childRect, parentRect) {
```
Comment (cbt1) [inline packages/picasso.js/src/core/math/intersection.js]: Not sure what you are trying to achieve here? parentRect doesn't contain any x or y

File: packages/picasso.js/src/core/math/intersection.js [hunk 6]
```diff
@@ -17,3 +32,40 @@ export function isCollinear(p1, p2, p3, e = EPSILON) {
...
+/**
+ * Check if the child rect is located entirely within it's parent rect. If no settings object is provided,
+ * it tries to check if the child rect fits assuming ltr mode and orientation at middle of the child rect.
+ * Returns true if child rect fits inside the parent rect, otherwise false.
+ * @param {rect} childRect - A rect describing the position(x,y), width and height of label
+ * @param {rect} parentRect - A rect describing the position(x,y), width and height of parent rect
+ * @param {object} settings - An object containing alignment(rtl or ltr) of childRect and orientation (optional)
+ * @ignore
+ */
+export function rectContainsRect(childRect, parentRect, settings) {
```
Comment (cbt1) [inline packages/picasso.js/src/core/math/intersection.js]: Please move this function and the other utility functions back to the labels namespace. As this is not a generic function for rectContainsRect. That would look something like this:
```
function rectContainsRect(a, b) {
    return a.x >= b.x &&
        a.x + a.width <= b.x + b.width &&
        a.y >= b.y &&
        a.y + a.height <= b.y + b.height;
}
```

Also I wouldn't bother with the type checking and throwing errors. As we don't have that else where in picasso.

File: packages/picasso.js/src/core/math/intersection.js [hunk 7]
```diff
@@ -17,3 +17,10 @@ export function isCollinear(p1, p2, p3, e = EPSILON) {
   const t = Math.abs((p1.x * (p2.y - p3.y)) + (p2.x * (p3.y - p1.y)) + (p3.x * (p1.y - p2.y)));
   return t < e;
 }
+
+export function rectContainsRect(a, b) {
```
Comment (cbt1) [inline packages/picasso.js/src/core/math/intersection.js]: Please add docs :)

**Comment by cbt1** [comment]
You may want to add `fix(labels)` to your squash and merge commit title

**Pull request merged**